File: src/main/python/coderuns2/lucky_number.py
import io
import os
import sys


def codeforces.round995.codeforces.round995.codeforces.round995.codeforces.round995.main():
    optim_input = io.BytesIO(os.read(0, os.fstat(0).st_size)).readline
    str_num = optim_input().decode().strip()
    if str_num == '0':
        return '1'
    sys.stdout.write(f'{nextLucky(str_num)}\n')


def nextLucky(strNum):
    n = len(strNum) // 2

    left = strNum[:n]
    right = strNum[n:]

    if n == 1:
        lf, r = int(left), int(right)
        if lf == 9 and r == 9:
            return '11'
        elif lf <= r:
            return str(lf + 1) + str(lf + 1)
        return left + left

    sl = dsum(left)
    sr = dsum(right)

    if sl <= sr:
        new_right = next_for_lower_sum(right, sl)
        if len(new_right) > n:
            new_left = str(increment(left))
            if len(new_left) > n:
                zeros = ''.join(['0' for _ in range(n - 1)])
                return f'{zeros}1{zeros}1'
            zeros = ''.join(['0' for _ in range(n)])
            new_right = next_for_higher_sum(zeros, dsum(new_left), n)
            return add_zeros(new_left, n) + new_right
        return left + new_right
    else:
        return left + next_for_higher_sum(right, sl, n)


def next_for_lower_sum(strNum, zum):
    i = 1
    while zum > dsum(strNum[:i]):
        i += 1
    i -= 1
    zeros = ['0' for _ in range(max(len(strNum) - i, 0))]
    front = f'{str(increment(strNum[:i]) if i > 0 else 1)}{"".join(zeros)}'
    rem = zum - dsum(front)
    end = add_zeros(minimize(rem), len(front))
    res = []
    for i in range(len(front)):
        res.append(str(int(front[i]) + int(end[i])))
    return ''.join(res)


def next_for_higher_sum(strNum, zum, n):
    diff = zum - dsum(strNum)
    i = 1
    while diff > i * 9 - dsum(strNum[-i:]):
        i += 1
    if i == 1:
        return add_zeros(increment(strNum, diff), n)
    suffix_sum = diff + dsum(strNum[-i:])
    return f'{strNum[:-i]}{minimize(suffix_sum)}'


def dsum(strNum):
    return sum([int(d) for d in strNum])


def minimize(rem) -> str:
    k = rem // 9
    m = rem % 9
    nines = ['9' for _ in range(k)]
    return f'{(str(m) if m > 0 else "")}{"".join(nines)}'


def add_zeros(strNum: str, length: int) -> str:
    if length - len(strNum) <= 0:
        return strNum
    zeros = ['0' for _ in range(length - len(strNum))]
    return f'{"".join(zeros)}{strNum}'


def increment(strNum: str, times: int = 1) -> str:
    increase = int(strNum[-1]) + times
    if increase // 10 == 0:
        return f'{strNum[:-1]}{str(increase)}'

    res = [str(increase % 10)]
    i = 2
    while i <= len(strNum) and strNum[-i] == '9':
        res.append('0')
        i += 1

    if i > len(strNum):
        res.append('1')
    else:
        res.append(str(int(strNum[-i]) + 1))
        i += 1
        while i <= len(strNum):
            res.append(strNum[-i])
            i += 1
    res.reverse()
    return ''.join(res)


def codeforces.round995.codeforces.round995.test():
    for max_length in range(1, 4):
        for ln in range(pow(10, max_length)):
            for p in range(len(str(ln)), max_length + 1):
                left = add_zeros(str(ln), max_length)
                for r in range(pow(10, max_length)):
                    right = add_zeros(str(r), max_length)
                    str_num = left + right
                    # print(str_num)
                    if brute_force(str_num) != nextLucky(str_num):
                        print(str_num, brute_force(str_num), '!=', nextLucky(str_num))
                        input()


def brute_force(str_num: str):
    n = len(str_num)
    num = int(str_num) + 1
    str_num = add_zeros(str(num), n)
    h = n // 2
    while dsum(str_num[:h]) != dsum(str_num[h:]):
        num += 1
        str_num = add_zeros(str(num), n)
    return add_zeros(str_num, n) if len(str_num) <= n else ('0' * (n // 2 - 1) + '1') * 2


if __name__ == '__main__':
    codeforces.round995.codeforces.round995.test()
    codeforces.round995.codeforces.round995.codeforces.round995.codeforces.round995.main()

File: coderun/src/main/kotlin/seasons/v2/LuckyNumber.kt
package seasons.v2

import java.io.BufferedReader
import java.io.InputStreamReader

fun main() {
//    test()

    val input = BufferedReader(InputStreamReader(System.`in`)).readLine()
    val strNum = input.trim()
    if (strNum == "0") {
        println("1")
    } else {
        println(nextLucky(strNum))
    }
}

fun nextLucky(strNum: String): String {
    val n = strNum.length / 2

    val left = strNum.substring(0, n)
    val right = strNum.substring(n)

    if (n == 1) {
        val l = left.toInt()
        val r = right.toInt()
        return when {
            l == 9 && r == 9 -> "11"
            l <= r -> "${l + 1}${l + 1}"
            else -> left + left
        }
    }

    val sl = dsum(left)
    val sr = dsum(right)

    return if (sl <= sr) {
        var newRight = nextForLowerSum(right, sl, n)
        if (newRight.length > n) {
            val newLeft = increment(left)
            if (newLeft.length > n) {
                val zeros = "0".repeat(n - 1)
                return "${zeros}1${zeros}1"
            }
            val zeros = "0".repeat(n)
            newRight = nextForHigherSum(zeros, dsum(newLeft), n)
            return addZeros(newLeft, n) + newRight
        }
        left + newRight
    } else {
        left + nextForHigherSum(right, sl, n)
    }
}

fun nextForLowerSum(strNum: String, sum: Int, n: Int): String {
    var i = 1
    while (sum > dsum(strNum.substring(0, i))) {
        i++
    }
    i--
    val zeros = "0".repeat(maxOf(strNum.length - i, 0))
    val front = "${if (i > 0) increment(strNum.substring(0, i)) else 1}$zeros"
    val rem = sum - dsum(front)
    val end = addZeros(minimize(rem), front.length)
    val res = StringBuilder()
    for (j in front.indices) {
        res.append((front[j].digitToInt() + end[j].digitToInt()).toString())
    }
    return res.toString()
}

fun nextForHigherSum(strNum: String, sum: Int, n: Int): String {
    val diff = sum - dsum(strNum)
    var i = 1
    while (diff > i * 9 - dsum(strNum.takeLast(i))) {
        i++
    }
    return if (i == 1) {
        addZeros(increment(strNum, diff), n)
    } else {
        val suffixSum = diff + dsum(strNum.takeLast(i))
        strNum.dropLast(i) + minimize(suffixSum)
    }
}

fun dsum(strNum: String): Int {
    return strNum.sumOf { it.digitToInt() }
}

fun minimize(rem: Int): String {
    val k = rem / 9
    val m = rem % 9
    val nines = "9".repeat(k)
    return if (m > 0) "$m$nines" else nines
}

fun addZeros(strNum: String, length: Int): String {
    return if (length - strNum.length <= 0) {
        strNum
    } else {
        "0".repeat(length - strNum.length) + strNum
    }
}

fun increment(strNum: String, times: Int = 1): String {
    val increase = strNum.last().digitToInt() + times
    if (increase < 10) {
        return strNum.dropLast(1) + increase.toString()
    }

    val res = StringBuilder()
    res.append((increase % 10).toString())
    var i = 2
    while (i <= strNum.length && strNum[strNum.length - i] == '9') {
        res.append('0')
        i++
    }

    if (i > strNum.length) {
        res.append('1')
    } else {
        res.append((strNum[strNum.length - i].digitToInt() + 1).toString())
        i++
        while (i <= strNum.length) {
            res.append(strNum[strNum.length - i])
            i++
        }
    }
    return res.reverse().toString()
}

fun bruteForce(strNum: String): String {
    var num = strNum.toInt() + 1
    val n = strNum.length
    var strNumStr = addZeros(num.toString(), n)
    val h = n / 2
    while (dsum(strNumStr.substring(0, h)) != dsum(strNumStr.substring(h))) {
        num++
        strNumStr = addZeros(num.toString(), n)
    }
    return if (strNumStr.length <= n) {
        addZeros(strNumStr, n)
    } else {
        "0".repeat(n / 2 - 1) + "1" + "0".repeat(n / 2 - 1) + "1"
    }
}

private fun test() {
    for (maxLength in 1..3) {
        for (l in 0 until Math.pow(10.0, maxLength.toDouble()).toInt()) {
            for (p in l.toString().length..maxLength) {
                val left = addZeros(l.toString(), maxLength)
                for (r in 0 until Math.pow(10.0, maxLength.toDouble()).toInt()) {
                    val right = addZeros(r.toString(), maxLength)
                    val strNum = left + right
                    if (bruteForce(strNum) != nextLucky(strNum)) {
                        println("$strNum ${bruteForce(strNum)} != ${nextLucky(strNum)}")
                    }
                }
            }
        }
    }
}
